lib/remap-lang/grammar.pest

```diff
@@ -0,0 +1,80 @@
+// Root ------------------------------------------------------------------------
+
+program    = _{ SOI ~ NEWLINE* ~ (expression ~ (NEWLINE+ ~ expression)*)* ~ NEWLINE* ~ EOI }
+expression = _{ statement | boolean_expr | block }
+statement  = _{ assignment | if_statement }
+
+// Statements ------------------------------------------------------------------
+
+assignment   = { target ~ "=" ~ expression }
+if_statement = { "if" ~ (boolean_expr | group) ~ block ~ ("else" ~ block)? }
+
+// Primary ---------------------------------------------------------------------
+
+primary  =  { value | variable | path | group }
+value    =  { string | float | integer | boolean | null }
+variable = ${ "$" ~ ident }
+group    =  { "(" ~ expression ~ ")" }
+
+// Function Calls --------------------------------------------------------------
+
+call      = ${ ident ~ "(" ~ arguments? ~ ")"  }
+arguments = !{ argument ~ ("," ~ argument)* }
+argument  =  { (ident ~ "=")? ~ (expression | regex) }
+
+// Operations ------------------------------------------------------------------
+
+boolean_expr   = { equality ~ (operator_boolean_expr ~ equality)* }
+equality       = { comparison ~ (operator_equality ~ comparison)* }
+comparison     = { addition ~ (operator_comparison ~ addition)* }
+addition       = { multiplication ~ (operator_addition ~ multiplication)* }
+multiplication = { not ~ (operator_multiplication ~ not)* }
+not            = { operator_not* ~ (call | primary) }
+
+// Operators -------------------------------------------------------------------
+
+operator_boolean_expr   = { "||" | "&&" }
```

Comment:
Right. That should already be the case, as the precedence is determined by the parser (see `operation_fns`). I'm not sure what the right test to prove this would be, though, since both `(true || false) && true` and `true || (false && true)` would evaluate to the same result.

I'll update the order of this rule to make it more explicit in the grammar.

However, I just noticed there's a bug in the implementation, in that both sides of the operator are evaluated _before_ the operation is applied, which means the code doesn't short-circuit (as one would expect, I assume).

This is already the case in the current implementation, so I'll leave it as is in this PR, but I'm going to create an issue to see what the concensus is on this.

Reply:
Filed as https://github.com/timberio/vector/issues/4706.